### PR TITLE
OR Query

### PR DESCRIPTION
**Pull request by tom-andersen**

### CODE REVIEW

File: firestore/src/Filter.cs [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    /// <summary>
+    /// A Filter represents a restriction on one or more field values and can be used to refine the results of a Query.
+    /// </summary>
+    public sealed class Filter {
+        private Filter(FilterProxy proxy) {
+            Proxy = proxy;
+        }
+        
+        internal FilterProxy Proxy { get; }
+
```
Comment (ehsannas) [inline firestore/src/Filter.cs:29]: Nit: It'd be more consistent with other sources to do:

```c#
private readonly FilterProxy _proxy;

private Filter(FilterProxy proxy) {
  _proxy = Util.NotNull(proxy);
}

internal FilterProxy Proxy {
  get {
    return _proxy;
  }
}
```

File: firestore/src/Filter.cs [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+            var filterVector = new FilterVector();
+            foreach (Filter filter in filters) {
+                filterVector.PushBack(filter.Proxy);
+            }
+            return new Filter(FirestoreCpp.FilterAnd(filterVector));
+        }
+
+        /// <summary>
+        /// Creates a new filter that is a disjunction of the given filters. A disjunction filter includes a document
+        /// if it satisfies <em>any</em> of the given filters.
```
Comment (ehsannas) [inline firestore/src/Filter.cs:298]: nit: I'm not sure if `<em>` and `</em>` work.

File: firestore/testapp/Assets/Firebase/Sample/Firestore/UIHandlerAutomated.cs [hunk 3]
```diff
@@ -2461,6 +2463,168 @@ private class TestException : Exception {
...
+        }));
+
+        AssertQueryResults(
+          desc: "Empty And",
+          query: c.Where(Filter.And()),
+          docIds: AsList("a", "b", "c"));
+
+        AssertQueryResults(
+          desc: "Empty Or",
+          query: c.Where(Filter.And()),
```
Comment (ehsannas) [inline firestore/testapp/Assets/Firebase/Sample/Firestore/UIHandlerAutomated.cs:2599]: should be empty Or()

**Review by ehsannas — COMMENTED** [review]
Looks good! thank you. Just a few nits pointed out below.

File: firestore/testapp/Assets/Firebase/Sample/Firestore/UIHandlerAutomated.cs [hunk 4]
```diff
@@ -2461,6 +2463,168 @@ private class TestException : Exception {
...
+            Filter.EqualTo("active", false))),
+          docIds: AsList("b"));
+
+        AssertQueryResults(
+          desc: "GreaterThan OR LessThan",
+          query: c.Where(Filter.Or(
+            Filter.GreaterThan("num", 2),
+            Filter.LessThan("num", 2))),
+          docIds: AsList("a", "c"));
+      });
```
Comment (ehsannas) [inline firestore/testapp/Assets/Firebase/Sample/Firestore/UIHandlerAutomated.cs:2625]: Please add two tests for nested cases:
and(a, or(b,c))
or(a, and(b,c))

File: firestore/testapp/Assets/Firebase/Sample/Firestore/UIHandlerAutomated.cs [hunk 5]
```diff
@@ -155,6 +155,8 @@ public class UIHandlerAutomated : UIHandler {
         TestInvalidArgumentAssertions,
         TestFirestoreDispose,
         TestFirestoreGetInstance,
+        TestWhereFilterQueries,
+        TestAndOrQueriesNoCompositeIndexes
```
Comment (ehsannas) [inline firestore/testapp/Assets/Firebase/Sample/Firestore/UIHandlerAutomated.cs:159]: It looks like `TestAndOrQueriesNoCompositeIndexes` tests accidentally got deleted?

File: firestore/testapp/Assets/Firebase/Sample/Firestore/UIHandlerAutomated.cs [hunk 6]
```diff
@@ -168,6 +170,7 @@ public class UIHandlerAutomated : UIHandler {
         TestTerminateMultiDBIndependently,
         TestTerminateAppWithMultiDB,
         TestRestartCustomFirestore,
+        TestAndOrQueriesCompositeIndexesRequired
```
Comment (ehsannas) [inline firestore/testapp/Assets/Firebase/Sample/Firestore/UIHandlerAutomated.cs:173]: It looks like `TestAndOrQueriesCompositeIndexesRequired` tests accidentally got deleted?

File: firestore/testapp/Assets/Firebase/Sample/Firestore/UIHandlerAutomated.cs (hunk 5 above)
Comment (tom-andersen) [reply firestore/testapp/Assets/Firebase/Sample/Firestore/UIHandlerAutomated.cs:159]: Looks like I messed up resolving merge conflicts. It should be back now.

**Review by ehsannas — APPROVED** [review]

**Pull request merged**